pymongo/auth_aws.py

```diff
@@ -88,7 +96,13 @@ def _authenticate_aws(credentials, sock_info):
                 # SASL complete.
                 break
     except PyMongoAuthAwsError as exc:
+        # Clear the cached credentials if we hit a failure in auth.
+        set_cached_credentials(None)
         # Convert to OperationFailure and include pymongo-auth-aws version.
         raise OperationFailure(
             "%s (pymongo-auth-aws version %s)" % (exc, pymongo_auth_aws.__version__)
         )
+    except Exception:
+        # Clear the cached credentials if we hit a failure in auth.
+        set_cached_credentials(None)
+        raise
```

Comment:
Can you confirm that these `set_cached_credentials(None)` calls are not strictly needed? Otherwise we would be breaking PyMongo<4.3

Reply:
Should we consider adding `pymongo_auth_aws.enable_caching(True)` to make it an opt in?

Reply:
Done: https://github.com/mongodb/pymongo-auth-aws/pull/11

Reply:
One case this ignores is when an app uses both explicit and implicit aws creds for two different clients. For example, let's say an app:

1. creates a client with implicit creds and the temp creds are cached
1. then creates a new client with invalid explicit creds.
1. The invalid creds will cause an auth failure and will clear the cached temp creds (from 1) which are actually still valid.

Do you think this case is worth handling? One fix might be to pass the AwsCredential to a new function `clear_cache(creds)`. Then the lib can determine whether the fail auth attempt is caused by the cached creds or not.

Reply:
I intentionally wanted to keep the cache as simple as possible, since rate limiting is hard to hit in practice.

Reply:
SGTM